File: scripts/producer/guided_presenter_profile.py
"""Explicit presenter metadata classes, not activated production dispatch.

The owner still authenticates actual V8 request/policy and V2 bindings, acquires
selected media, proves original-prefix pixels and separately screens captions.
These helpers never strip a plan into a legacy profile or grant delivery QC.
"""
from __future__ import annotations

from fractions import Fraction

from audio.program_finish_contract import finishing_reason
from graphics.presenter_layout_contract import PresenterCanvas
from guided_caption_layers import caption_page_bound
from guided_caption_profile import caption_preset_plan
from guided_media_profile import manual_short_geometry
from guided_presenter_assets import _windows
from guided_presenter_probe_contract import canonical_presenter_rate
from opening_prefix_contract import GRAPH_WORKLOAD_POLICY, MAX_CLIPS, PrefixClock, valid_canvas

PRESENTER_PROFILE = "unity-source-float-own-screen-presenter-layout-v1"
PRESENTER_SHORT_PROFILE = "unity-source-float-manual-short-own-screen-presenter-layout-v1"
PRESENTER_CAPTION_PROFILE = "unity-source-float-own-screen-presenter-caption-layout-v1"
PRESENTER_CAPTION_SHORT_PROFILE = "unity-source-float-manual-short-own-screen-presenter-caption-layout-v1"
PRESENTER_BODY_PROFILE = "held-source-float-own-screen-presenter-layout-body-v1"
PRESENTER_SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-presenter-layout-body-v1"
PRESENTER_CAPTION_BODY_PROFILE = "held-source-float-own-screen-presenter-caption-layout-body-v1"
PRESENTER_CAPTION_SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-presenter-caption-layout-body-v1"
_BODY = {PRESENTER_PROFILE: PRESENTER_BODY_PROFILE, PRESENTER_SHORT_PROFILE: PRESENTER_SHORT_BODY_PROFILE,
         PRESENTER_CAPTION_PROFILE: PRESENTER_CAPTION_BODY_PROFILE,
         PRESENTER_CAPTION_SHORT_PROFILE: PRESENTER_CAPTION_SHORT_BODY_PROFILE}


def presenter_opening_profile(value: object) -> str:
    """Read only a distinct new token; historical tokens never change their meaning."""
    if type(value) is not str or value not in _BODY:
        raise RuntimeError("presenter opening profile is unsupported")
    return value


def presenter_body_profile(value: object) -> str:
    """Keep exact one-to-one opening/body pairing, never upgrade an old result."""
    return _BODY[presenter_opening_profile(value)]


def presenter_caption_profile(value: object) -> bool:
    """A captioned presenter needs both existing graphic and new subject clearance."""
    return type(value) is str and value in (PRESENTER_CAPTION_PROFILE, PRESENTER_CAPTION_SHORT_PROFILE)


def presenter_manual_profile(value: object) -> bool:
    """This class accepts explicit geometry, not new tracking or crop authoring."""
    return type(value) is str and value in (PRESENTER_SHORT_PROFILE, PRESENTER_CAPTION_SHORT_PROFILE)


def _canvas(plan: dict, clock: PrefixClock) -> bool:
    """Require exact native destination and bounded actual clock metadata."""
    canonical_presenter_rate(clock.frame_rate)
    rate = Fraction(clock.frame_rate)
    if clock.frame_rate != f"{rate.numerator}/{rate.denominator}" or not 1 <= rate <= 60 \
            or type(clock.total_frames) is not int or not 1 <= clock.total_frames <= 72000:
        raise RuntimeError("presenter profile requires the existing bounded exact guided clock")
    target = plan.get("target")
    if type(target) is not dict or target.get("mode") not in ("longform", "short"):
        raise RuntimeError("presenter needs an explicit longform/manual-short target")
    short = target["mode"] == "short"
    dimensions = (1080, 1920) if short else (1920, 1080)
    if any(type(target.get(key)) is not int for key in ("width", "height")) \
            or (target["width"], target["height"]) != dimensions or (clock.width, clock.height) != dimensions:
        raise RuntimeError("presenter needs the exact native1080 destination")
    _windows(plan.get("presenterLayouts"), PresenterCanvas(*dimensions, clock.total_frames, "yuv420p"))
    return short


def presenter_profile_for_plan(plan: dict, clock: PrefixClock) -> str:
    """Validate nonempty declared geometry/preset without selecting a live workflow."""
    if type(plan) is not dict or type(clock) is not PrefixClock:
        raise RuntimeError("presenter profile requires exact plan and whole-clock metadata")
    short = _canvas(plan, clock)
    if any(key in plan for key in ("presenter", "overlays")):
        raise RuntimeError("presenter profile has inherited incompatible visual intent")
    unsupported = ("titleCards", "brollTrack", "baselineLook", "punchIns", "transitions",
                   "persistentText", "faceBBoxNorm", "chapters")
    if any(plan.get(key) for key in unsupported):
        raise RuntimeError("presenter profile has additional unqualified finishing lanes")
    # audioEnhance/audioGain finish in the shared program master (audio/program_finish_bus);
    # they are validated explicitly here, never dropped and never accepted malformed.
    finishing = finishing_reason(plan)
    if finishing:
        raise RuntimeError("presenter profile rejects " + finishing)
    if any(key in plan for key in ("captionCorrectionLedger", "captionStyles", "captionChapters", "dialogueCaptionAuthority")):
        raise RuntimeError("presenter profile has unqualified alternate caption authority")
    cuts = plan.get("cutTrack")
    if type(cuts) is not list or not cuts or any(type(row) is not dict or type(row.get("speed", 1)) not in (int, float)
            or row.get("speed", 1) != 1 or type(row.get("audioLeadMs", 0)) not in (int, float)
            or row.get("audioLeadMs", 0) != 0 for row in cuts):
        raise RuntimeError("presenter profile requires unchanged speed-one cuts without J-cut leads")
    captioned = "captionsTrack" in plan
    if captioned:
        caption_preset_plan(plan)
    elif type(plan.get("captions")) is not dict or set(plan["captions"]) != {"burn"} \
            or plan["captions"]["burn"] is not False:
        raise RuntimeError("uncaptioned presenter requires explicit captions.burn=false")
    if short:
        manual_short_geometry(plan, captioned)
        return PRESENTER_CAPTION_SHORT_PROFILE if captioned else PRESENTER_SHORT_PROFILE
    if plan.get("reframe") not in (None, {}, {"strategy": "none"}):
        raise RuntimeError("longform presenter has unqualified reframe intent")
    return PRESENTER_CAPTION_PROFILE if captioned else PRESENTER_PROFILE


def presenter_graph_workload(clock: PrefixClock, graphic_count: int, captioned: bool,
                             asset_canvases: tuple[tuple[int, int], ...]) -> dict:
    """Count every prospective native input occurrence, not only unique file paths."""
    if type(clock) is not PrefixClock or not valid_canvas(clock.width, clock.height) \
            or type(graphic_count) is not int or not 0 <= graphic_count <= 128 or type(captioned) is not bool:
        raise RuntimeError("presenter graph workload metadata is malformed")
    canonical_presenter_rate(clock.frame_rate)
    PresenterCanvas(clock.width, clock.height, clock.total_frames, "yuv420p")
    if type(asset_canvases) is not tuple or not 1 <= len(asset_canvases) <= 32 or any(
            type(row) is not tuple or len(row) != 2 or not valid_canvas(*row) for row in asset_canvases):
        raise RuntimeError("presenter graph needs1–32 bounded asset-canvas occurrences")
    pages = caption_page_bound(clock, (graphic_count, graphic_count), clock.total_frames)["fullPageBound"] if captioned else 0
    count = graphic_count + pages + len(asset_canvases)
    pixels = (1 + graphic_count + pages) * clock.width * clock.height + sum(width * height for width, height in asset_canvases)
    if count > MAX_CLIPS or pixels > GRAPH_WORKLOAD_POLICY["maxGraphInputPixels"]:
        raise RuntimeError("presenter combined graph exceeds existing workload; no requested inputs may be dropped")
    return {"kind": "presenter-combined-workload-preflight", "scope": "metadata-not-observed-media-or-RSS-proof",
        "graphicCount": graphic_count, "captionPageBound": pages, "presenterOccurrences": len(asset_canvases),
        "fullGraphInputPixels": pixels, "productionPerformanceQualified": False}


def _span(row: dict, total: int) -> tuple[int, int]:
    """Use exact half-open frames, including the full authored ramp endpoints."""
    if type(row) is not dict:
        raise RuntimeError("presenter coexistence row is malformed")
    start, end = row.get("startFrame"), row.get("endFrameExclusive")
    if type(start) is not int or type(end) is not int or not 0 <= start < end <= total:
        raise RuntimeError("presenter coexistence requires bounded original frame windows")
    return start, end


def assert_presenter_graphics_disjoint(graphics: list[dict], windows: list[dict], total: int) -> None:
    """An opaque own-screen graphic cannot cover an explicitly requested presenter."""
    if type(total) is not int or total <= 0 or type(graphics) is not list or len(graphics) > 128 \
            or type(windows) is not list or not 1 <= len(windows) <= 32:
        raise RuntimeError("presenter coexistence requires the bounded whole-candidate inventory")
    graphic_spans = [_span(row, total) for row in graphics]
    presenter_spans = [_span(row, total) for row in windows]
    if any(max(a, c) < min(b, d) for a, b in graphic_spans for c, d in presenter_spans):
        raise RuntimeError("own-screen graphic overlaps requested presenter; no automatic retiming or suppression")

File: src/lib/producer/contracts/guided-presenter-profile.ts
/** Explicit, unreleased presenter classes. Metadata eligibility is not live ownership, clearance or approval. */
import { objectValue, exactKeys } from "./validation";
import { assertManualShortGeometry } from "./guided-media-profile";
import { assertCaptionPresetPlan } from "./guided-caption-profile";
import { parseTreatmentProposalV8, proposalV8ValidationView, type TreatmentProposalV8 } from "./treatment-proposal-v8";
import { proposalV7ValidationView } from "./treatment-proposal-v7";
import { proposalV6ValidationView } from "./treatment-proposal-v6";
import { assertPresenterGraphWorkload, presenterNativeCanvas } from "./guided-presenter-workload";
import { canonicalJsonSha256 } from "@/lib/server/auto-edit-hash";
import { applyGuidedPresenterOperations, assertGuidedPresenterCandidate, guidedPresenterPolicy,
  guidedPresenterWindows, type GuidedPresenterWindow } from "@/lib/server/guided-proposal-presenter";
import { applyGuidedMusicOperations } from "@/lib/server/guided-proposal-music";
import { applyGuidedCaptionOperations, guidedCaptionPolicy } from "@/lib/server/guided-proposal-captions";
import { assertGuidedReframeCandidate } from "@/lib/server/guided-proposal-reframe";
import { assertGuidedFrameBindings, type GuidedGraphicFrameBinding } from "@/lib/server/guided-proposal-bindings";
import type { ProposalEvidence } from "@/lib/server/guided-proposal-evidence";
export { assertPresenterGraphWorkload } from "./guided-presenter-workload";

export const PRESENTER_PROFILE = "unity-source-float-own-screen-presenter-layout-v1" as const;
export const PRESENTER_SHORT_PROFILE = "unity-source-float-manual-short-own-screen-presenter-layout-v1" as const;
export const PRESENTER_CAPTION_PROFILE = "unity-source-float-own-screen-presenter-caption-layout-v1" as const;
export const PRESENTER_CAPTION_SHORT_PROFILE = "unity-source-float-manual-short-own-screen-presenter-caption-layout-v1" as const;
export const PRESENTER_BODY_PROFILE = "held-source-float-own-screen-presenter-layout-body-v1" as const;
export const PRESENTER_SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-presenter-layout-body-v1" as const;
export const PRESENTER_CAPTION_BODY_PROFILE = "held-source-float-own-screen-presenter-caption-layout-body-v1" as const;
export const PRESENTER_CAPTION_SHORT_BODY_PROFILE = "held-source-float-manual-short-own-screen-presenter-caption-layout-body-v1" as const;
const BODY_BY_OPENING = { [PRESENTER_PROFILE]: PRESENTER_BODY_PROFILE, [PRESENTER_SHORT_PROFILE]: PRESENTER_SHORT_BODY_PROFILE,
  [PRESENTER_CAPTION_PROFILE]: PRESENTER_CAPTION_BODY_PROFILE, [PRESENTER_CAPTION_SHORT_PROFILE]: PRESENTER_CAPTION_SHORT_BODY_PROFILE } as const;
export type PresenterOpeningProfile = keyof typeof BODY_BY_OPENING;
export type PresenterBodyProfile = typeof BODY_BY_OPENING[PresenterOpeningProfile];
export interface GuidedPresenterProfileInput {
  accepted: Record<string, unknown>; candidate: Record<string, unknown>; manifest: Record<string, unknown>;
  proposal: unknown; evidence: ProposalEvidence; bindings: unknown;
}

/** Token classification only; it cannot select a profile for a plan or authenticate an execution. */
export function isPresenterOpeningProfile(value: unknown): value is PresenterOpeningProfile {
  return typeof value === "string" && Object.hasOwn(BODY_BY_OPENING, value);
}

export function isPresenterBodyProfile(value: unknown): value is PresenterBodyProfile {
  return typeof value === "string" && Object.values(BODY_BY_OPENING).some(profile => profile === value);
}

export function presenterOpeningProfile(value: unknown): PresenterOpeningProfile {
  if (typeof value !== "string" || !Object.hasOwn(BODY_BY_OPENING, value)) throw new Error("Unsupported explicit presenter opening profile");
  return value as PresenterOpeningProfile;
}

export function presenterBodyProfile(value: PresenterOpeningProfile): PresenterBodyProfile {
  return BODY_BY_OPENING[presenterOpeningProfile(value)];
}

export function isPresenterCaptionProfile(value: unknown): boolean {
  return value === PRESENTER_CAPTION_PROFILE || value === PRESENTER_CAPTION_SHORT_PROFILE;
}

export function isPresenterManualProfile(value: unknown): boolean {
  return value === PRESENTER_SHORT_PROFILE || value === PRESENTER_CAPTION_SHORT_PROFILE;
}

function active(value: unknown): boolean {
  if (Array.isArray(value)) return value.length > 0;
  if (value !== null && typeof value === "object") return Object.keys(value).length > 0;
  return Boolean(value);
}

function sameFields(expected: Record<string, unknown>, candidate: Record<string, unknown>, keys: string[]): void {
  if (keys.some(key => Object.hasOwn(candidate, key) !== Object.hasOwn(expected, key)
      || Object.hasOwn(expected, key) && canonicalJsonSha256(candidate[key]) !== canonicalJsonSha256(expected[key]))) {
    throw new Error("Presenter candidate prior lanes differ from actual requested or inherited fields");
  }
}

/** Views are local prior-lane validation only; actual V8 evidence/bindings/operation indices remain unchanged. */
function assertPriorLanes(input: GuidedPresenterProfileInput, proposal: TreatmentProposalV8): void {
  const musicView = proposalV8ValidationView(proposal), cropView = proposalV7ValidationView(musicView);
  const captionPolicy = input.evidence.captionPolicy;
  if (!captionPolicy || canonicalJsonSha256(captionPolicy) !== canonicalJsonSha256(guidedCaptionPolicy(captionPolicy.configuration))) {
    throw new Error("Presenter V8 evidence lacks the exact captured caption policy");
  }
  if (!input.evidence.musicPolicy) throw new Error("Presenter V8 evidence lacks its original music policy");
  const music = applyGuidedMusicOperations({ plan: input.accepted, manifest: input.manifest,
    proposal: musicView, policy: input.evidence.musicPolicy });
  sameFields(music, input.candidate, ["music"]);
  if (cropView.operations.some(row => row.type === "reframe-manual-short")) {
    assertGuidedReframeCandidate(input.accepted, input.candidate, cropView);
    return;
  }
  const captioned = applyGuidedCaptionOperations(input.accepted, proposalV6ValidationView(cropView));
  sameFields(captioned, input.candidate, ["reframe", "captions", "captionsTrack", "captionStyles", "captionCorrectionLedger",
    "captionChapters", "dialogueCaptionAuthority", "chapters"]);
}

function declaredCaptionMode(plan: Record<string, unknown>): boolean {
  const captions = objectValue(plan.captions, "presenter explicit caption choice");
  exactKeys(captions, ["burn"], ["burn"], "presenter explicit caption choice");
  if (captions.burn === true) { assertCaptionPresetPlan(plan); return true; }
  if (captions.burn !== false || Object.hasOwn(plan, "captionsTrack")) {
    throw new Error("Uncaptioned presenter requires explicit burn:false and absent captionsTrack");
  }
  return false;
}

function assertCandidateClass(plan: Record<string, unknown>, captioned: boolean): void {
  const target = objectValue(plan.target, "presenter target"); presenterNativeCanvas(target);
  const unsupported = ["titleCards", "brollTrack", "baselineLook", "punchIns", "transitions", "audioEnhance", "audioGain",
    "persistentText", "faceBBoxNorm", "chapters"];
  const authority = ["captionStyles", "captionCorrectionLedger", "captionChapters", "dialogueCaptionAuthority"];
  if (unsupported.some(key => active(plan[key])) || authority.some(key => Object.hasOwn(plan, key))) {
    throw new Error("Presenter profile has unqualified additional lane or caption authority");
  }
  if (target.mode === "short") assertManualShortGeometry(plan, captioned);
  else {
    const reframe = plan.reframe == null ? {} : objectValue(plan.reframe, "presenter longform reframe");
    if (Object.keys(reframe).some(key => key !== "strategy") || Object.hasOwn(reframe, "strategy") && reframe.strategy !== "none") {
      throw new Error("Presenter longform does not qualify reframing, tracking or implicit crop");
    }
  }
  if (!Array.isArray(plan.cutTrack) || !plan.cutTrack.length) throw new Error("Presenter needs nonempty unity cuts");
  for (const value of plan.cutTrack) {
    const cut = objectValue(value, "presenter cut");
    if (Object.hasOwn(cut, "speed") && cut.speed !== 1 || cut.audioLeadMs !== undefined && cut.audioLeadMs !== 0) {
      throw new Error("Presenter profile does not qualify speed changes or J-cut audio leads");
    }
  }
}

function assertGraphicClass(entry: Record<string, unknown>, row: GuidedGraphicFrameBinding): void {
  const p = row.presentation, spec = active(entry.spec) ? objectValue(entry.spec, "presenter graphic spec") : {};
  const hole = entry.kind === "module-takeover"
    || ["module-scoreboard", "module-pipeline", "module-ledger-dark"].includes(String(entry.kind)) && active(spec.presenterFrame);
  if (p.schemaVersion !== 1 || p.anchor !== "own-screen" || p.placement !== "full-canvas" || p.compositeMode !== "normal"
      || p.baseTreatment !== "preserve" || entry.anchor !== "own-screen" || entry.takeoverBase != null
      || active(entry.exitOnCut) || active(entry.placement) || hole) {
    throw new Error("Presenter profile requires normal full-canvas preserve graphics without holes or added placement");
  }
}

/** Whole-body half-open comparison; even all-future graphics cannot hide an explicitly requested presenter. */
export function assertPresenterGraphicsDisjoint(graphics: GuidedGraphicFrameBinding[], windows: GuidedPresenterWindow[]): void {
  if (graphics.some(graphic => windows.some(window => graphic.startFrame < window.endFrameExclusive
      && window.startFrame < graphic.endFrameExclusive))) {
    throw new Error("Presenter windows must not overlap any original full-canvas graphic; no inputs may be removed or retimed");
  }
}

function exactBindings(input: GuidedPresenterProfileInput, proposal: TreatmentProposalV8) {
  if (active(input.accepted.graphicsTrack)) throw new Error("Inherited graphics have no fresh exact presenter profile bindings");
  const bindings = assertGuidedFrameBindings({ proposal, evidence: input.evidence, candidate: input.candidate, inheritedCount: 0 }, input.bindings);
  if (bindings.schemaVersion !== 2 || !bindings.presenterLayouts.length || bindings.unboundInheritedGraphicIds.length) {
    throw new Error("Presenter profiles require nonempty actual V8 windows and exact V2 bindings");
  }
  const track = input.candidate.graphicsTrack as unknown[], ids = new Set<string>();
  for (const row of bindings.graphics) {
    const entry = objectValue(track[row.order], "presenter graphic");
    if (ids.has(row.graphicId) || !Number.isSafeInteger(row.startFrame) || !Number.isSafeInteger(row.endFrameExclusive)
        || row.startFrame < 0 || row.startFrame >= row.endFrameExclusive || row.endFrameExclusive > bindings.totalFrames) {
      throw new Error("Presenter graphic bindings lack exact ordered unique whole-body intervals");
    }
    ids.add(row.graphicId); assertGraphicClass(entry, row);
  }
  assertPresenterGraphicsDisjoint(bindings.graphics, bindings.presenterLayouts);
  return bindings;
}

function selectedProfile(short: boolean, captioned: boolean): PresenterOpeningProfile {
  if (captioned) return short ? PRESENTER_CAPTION_SHORT_PROFILE : PRESENTER_CAPTION_PROFILE;
  return short ? PRESENTER_SHORT_PROFILE : PRESENTER_PROFILE;
}

/** Pure metadata preflight only. Owners still owe original intent/source, live media, caption clearance and QC. */
export function assertGuidedPresenterProfile(input: GuidedPresenterProfileInput) {
  const proposal = parseTreatmentProposalV8(input.proposal);
  if (input.evidence.schemaVersion !== 8 || !input.evidence.presenterPolicy || proposal.colorPolicy !== "preserve"
      || proposal.operations.some(row => row.type === "grade")) throw new Error("Presenter profile requires actual V8 policy and unchanged source color");
  applyGuidedPresenterOperations({ plan: input.accepted, manifest: input.manifest, proposal,
    policy: input.evidence.presenterPolicy, evidence: input.evidence });
  assertGuidedPresenterCandidate({ ...input, proposal }); assertPriorLanes(input, proposal);
  const captioned = declaredCaptionMode(input.candidate); assertCandidateClass(input.candidate, captioned);
  const bindings = exactBindings(input, proposal), windows = guidedPresenterWindows(proposal, input.evidence);
  if (windows.some(row => row.layout.presenterRect.width > row.layout.presenterCrop.width
      || row.layout.presenterRect.height > row.layout.presenterCrop.height)) throw new Error("Initial presenter executor does not qualify presenter upscaling");
  const assets = guidedPresenterPolicy(input.accepted, input.manifest).assets;
  const canvases = windows.map(row => assets.find(asset => asset.assetId === row.layout.assetId)!);
  const workload = assertPresenterGraphWorkload({ target: input.candidate.target, clock: bindings,
    graphicCount: bindings.graphics.length, captioned, presenterAssetCanvases: canvases });
  const profile = selectedProfile(objectValue(input.candidate.target, "presenter target").mode === "short", captioned);
  return { profile, bodyProfile: presenterBodyProfile(profile), windows, workload };
}
